File: MCPForUnity/Editor/Tools/ValidateBuild.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEditor.Build;
using MCPForUnity.Editor.Helpers;
using MCPForUnity.Editor.Tools.Build;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// READ-only build pre-flight. Inspects whether a player build for a given
    /// target would succeed WITHOUT producing any artifacts or mutating project
    /// state. Safe to run in read_only/review_only before a real manage_build.
    /// </summary>
    [McpForUnityTool("validate_build", AutoRegister = false, Group = "core")]
    public static class ValidateBuild
    {
        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
                return new ErrorResponse("Parameters cannot be null.");

            try
            {
                var p = new ToolParams(@params);
                string targetName = p.Get("target");

                var checks = new List<object>();
                var issues = new List<string>();
                bool wouldBuild = true;

                // (a) target resolves — reuse ManageBuild's mapping
                if (!BuildTargetMapping.TryResolveBuildTarget(targetName, out var target))
                {
                    string detail = BuildTargetMapping.GetUnknownBuildTargetMessage(targetName);
                    checks.Add(Check("target_resolves", false, detail));
                    issues.Add(detail);
                    // Cannot proceed with further target-dependent checks.
                    return new SuccessResponse("Build pre-flight complete.", new
                    {
                        would_build = false,
                        target = targetName,
                        checks,
                        issues
                    });
                }
                checks.Add(Check("target_resolves", true, $"Resolved to {target}."));

                // (b) build support installed for this target
                var group = BuildTargetMapping.GetTargetGroup(target);
                bool supported = BuildPipeline.IsBuildTargetSupported(group, target);
                if (!supported)
                {
                    string detail = $"Build support for '{target}' is not installed. Install it via Unity Hub.";
                    checks.Add(Check("build_support_installed", false, detail));
                    issues.Add(detail);
                    wouldBuild = false;
                }
                else
                {
                    checks.Add(Check("build_support_installed", true, $"Build support for {target} is installed."));
                }

                // (c) scenes in EditorBuildSettings — count, enabled count, on-disk existence
                var allScenes = EditorBuildSettings.scenes;
                int totalScenes = allScenes.Length;
                var enabledScenes = allScenes.Where(s => s.enabled).ToArray();
                int enabledCount = enabledScenes.Length;

                var missingScenes = enabledScenes
                    .Where(s => string.IsNullOrEmpty(s.path) || !File.Exists(s.path))
                    .Select(s => s.path ?? "(empty path)")
                    .ToArray();

                if (enabledCount == 0)
                {
                    string detail = "No enabled scenes in Build Settings. The build would use only the active scene (if any).";
                    checks.Add(Check("scenes_enabled", false, detail));
                    issues.Add(detail);
                    // A scene-less build can still technically produce a player; do not
                    // fail wouldBuild solely on this, but surface it as an issue.
                }
                else if (missingScenes.Length > 0)
                {
                    string detail =
                        $"{missingScenes.Length} enabled scene(s) do not exist on disk: {string.Join(", ", missingScenes)}.";
                    checks.Add(Check("scenes_enabled", false, detail));
                    issues.Add(detail);
                    wouldBuild = false;
                }
                else
                {
                    checks.Add(Check("scenes_enabled", true,
                        $"{enabledCount} of {totalScenes} scene(s) enabled; all exist on disk."));
                }

                // (d) no pending compile errors
                bool compilationFailed = EditorUtility.scriptCompilationFailed;
                if (compilationFailed)
                {
                    string detail = "Scripts failed to compile — fix compile errors before building.";
                    checks.Add(Check("scripts_compile", false, detail));
                    issues.Add(detail);
                    wouldBuild = false;
                }
                else
                {
                    checks.Add(Check("scripts_compile", true, "No script compilation errors detected."));
                }

                // (e) PlayerSettings sanity — productName / companyName / applicationIdentifier
                var namedTarget = BuildTargetMapping.GetNamedBuildTarget(target);
                CheckNonEmptyPlayerSetting("product_name", PlayerSettings.productName, checks, issues, ref wouldBuild);
                CheckNonEmptyPlayerSetting("company_name", PlayerSettings.companyName, checks, issues, ref wouldBuild);

                string appId = SafeGetApplicationIdentifier(namedTarget);
                CheckNonEmptyPlayerSetting("application_identifier", appId, checks, issues, ref wouldBuild);

                return new SuccessResponse("Build pre-flight complete.", new
                {
                    would_build = wouldBuild,
                    target = target.ToString(),
                    checks,
                    issues
                });
            }
            catch (Exception ex)
            {
                return new ErrorResponse(ex.Message, new { stackTrace = ex.StackTrace });
            }
        }

        private static object Check(string name, bool ok, string detail)
        {
            return new { name, ok, detail };
        }

        private static void CheckNonEmptyPlayerSetting(
            string name, string value, List<object> checks, List<string> issues, ref bool wouldBuild)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                string detail = $"PlayerSettings.{name} is empty.";
                checks.Add(Check(name, false, detail));
                issues.Add(detail);
                wouldBuild = false;
            }
            else
            {
                checks.Add(Check(name, true, $"{name} = '{value}'."));
            }
        }

        private static string SafeGetApplicationIdentifier(NamedBuildTarget namedTarget)
        {
            try
            {
                return PlayerSettings.GetApplicationIdentifier(namedTarget);
            }
            catch
            {
                return null;
            }
        }
    }
}

File: Server/src/services/tools/validate_build.py
"""Read-only build pre-flight — would a build for target X succeed?

This inspects build settings and reports problems WITHOUT producing any
artifacts or mutating project state. Unlike manage_build(action=build) — which
is DESTRUCTIVE (runs build hooks, writes to disk) — validate_build is a pure
READ and is safe to run in read_only/review_only before a real build.
"""

from typing import Annotated, Any, Optional

from fastmcp import Context
from mcp.types import ToolAnnotations

from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry


@mcp_for_unity_tool(
    group="core",
    description=(
        "Read-only build pre-flight (READ; produces no artifacts, mutates nothing). "
        "Answers 'would a build for target X succeed?' before running the DESTRUCTIVE "
        "manage_build. Checks: the target resolves; its build support is installed; "
        "Build Settings scenes exist on disk; scripts compile; and PlayerSettings "
        "(productName / companyName / applicationIdentifier) are non-empty. Returns "
        "{would_build, target, checks: [{name, ok, detail}], issues: [...]}. "
        "Target defaults to the active build target when omitted."
    ),
    annotations=ToolAnnotations(
        title="Validate Build",
        readOnlyHint=True,
    ),
)
async def validate_build(
    ctx: Context,
    target: Annotated[
        Optional[str],
        "Build target to validate: windows64, osx, linux64, android, ios, webgl, "
        "uwp, tvos, visionos. Omit to validate the active build target.",
    ] = None,
) -> dict[str, Any]:
    params_dict: dict[str, Any] = {}
    if target is not None:
        params_dict["target"] = target

    unity_instance = await get_unity_instance_from_context(ctx)
    result = await send_with_unity_instance(
        async_send_command_with_retry, unity_instance, "validate_build", params_dict
    )
    return result if isinstance(result, dict) else {"success": False, "message": str(result)}
